website/registrations/apps.py

```diff
@@ -41,3 +43,59 @@ def user_menu_items(self):
                 },
             ],
         }
+
+    @staticmethod
+    def execute_data_minimisation(dry_run=False):
+        """Delete completed or rejected registrations that were modified at least 31 days ago.
+
+        :param dry_run: does not really remove data if True
+        :return: number of removed objects.
+        """
+        from .models import Entry, Registration, Renewal
+
+        deletion_period = timezone.now() - timezone.timedelta(days=31)
+        registrations = Registration.objects.filter(
+            Q(status=Entry.STATUS_COMPLETED) | Q(status=Entry.STATUS_REJECTED),
+            updated_at__lt=deletion_period,
+        )
+        renewals = Renewal.objects.filter(
+            Q(status=Entry.STATUS_COMPLETED) | Q(status=Entry.STATUS_REJECTED),
+            updated_at__lt=deletion_period,
+        )
+
+        if dry_run:
+            return registrations.count() + renewals.count()  # pragma: no cover
+
+        # Mark that this deletion is for data minimisation so that it can be recognized
+        # in any post_delete signal handlers. This is used to prevent the deletion of
+        # Moneybird invoices.
+        registrations = registrations.annotate(
+            __deleting_for_dataminimisation=Value(True)
+        )
+        renewals = renewals.annotate(__deleting_for_dataminimisation=Value(True))
+
+        return registrations.delete()[0] + renewals.delete()[0]
+
+    @staticmethod
+    def minimise_user(user, dry_run: bool = False) -> None:
+        """Minimise the data stored for a user."""
+        from .models import Entry, Renewal
+
+        renewals = Renewal.objects.filter(
+            member=user,
+        )
+
+        open_renewals = renewals.filter(
+            (
+                Q(status=Entry.STATUS_CONFIRM)
+                | Q(status=Entry.STATUS_ACCEPTED)
+                | Q(status=Entry.STATUS_REVIEW)
+            ),
+            member=user,
+        )
+        if open_renewals.exists():
+            raise ValueError("Cannot minimise user with open renewals.")
+
+        if not dry_run:
+            renewals.delete()
```

Comment:
Renewals already cascade on member deletion. As you'll delete them explicitly here, it can just get PROTECT. The payment should get PROTECT instead of SET_NULL.